### ISSUE TITLE
`{{component}}` can't be invoked directly

### ISSUE THREAD

**Issue by boris-petrov** [typed-ember/glint#84]
[This commit](https://github.com/boris-petrov/glint-issues/commit/8df69f14352ac0a1f736c1e49cb0435e689cf037) of my reproduction repo.

I think the `component` component isn't type-checked correctly at all. Pretty much all my invocations are marked as an error. As you can see in this example:

```
app/components/test-component.hbs:3:1 - error TS2345: Argument of type 'new () => Invokable<(args: Omit<{ index: number; }, "index"> & Partial<Pick<{ index: number; }, "index">>) => AcceptsBlocks<EmptyObject>>' is not assignable to parameter of type 'string | number | boolean | void | AcceptsBlocks<{}> | null'.
  Type 'new () => Invokable<(args: Omit<{ index: number; }, "index"> & Partial<Pick<{ index: number; }, "index">>) => AcceptsBlocks<EmptyObject>>' is not assignable to type 'AcceptsBlocks<{}>'.
    Type 'new () => Invokable<(args: Omit<{ index: number; }, "index"> & Partial<Pick<{ index: number; }, "index">>) => AcceptsBlocks<EmptyObject>>' provides no match for the signature '(blocks: {}): { [Blocks]: true; }'.

3 {{component "glimmer-component" index=1}}
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I guess that would be difficult to fix so that's one of the reasons I requested some [ignore syntax](https://github.com/typed-ember/glint/issues/82).

**Comment by dfreeman** [comment]
The `{{component}}` keyword _is_ hard to type robustly, but we've got it most of the way there. You've definitely found a pretty key bug, though 😄

As a workaround in the meantime, this should typecheck as expected:

```hbs
{{#let (component "glimmer-component" index=1) as |GlimmerComponent|}}
  <GlimmerComponent />
{{/let}}
```

**Comment by dfreeman** [comment]
I've spent a little bit of time with this, and unfortunately we might end up having to stick it in the `wontfix` bucket. Similar to `ObjectProxy`, it's something we could make work but at the expense of complicating the types (and therefore slowing down the checking) for every component/helper invocation.

Also similar to `ObjectProxy`, the framework is moving in a direction where this pattern should be needed less and less. With the changes that landed in Ember 3.25, places where you would have done something like this before:

```ts
class MyComponent ... {
  get whichThingToRender(): 'foo' | 'bar' {
    // ...
  }
}
```

```hbs
{{#component this.whichThingToRender}}
  {{! ... }}
{{/component}}
```

Can now become this:

```ts
import Foo from '...';
import Bar from '...';

class MyComponent ... {
  get whichThingToRender(): typeof Foo | typeof Bar {
    // ...
  }
}
```

```hbs
<this.whichThingToRender>
  {{! ... }}
</this.whichThingToRender>
```

Between the complexity and perf implications of making this work, the Ember 3.25 changes, and the `{{#let}}` workaround, I think this is unlikely to get fixed in the near term.

I'm happy to keep this open for now, though, and if you've got other places you're direct-invoking dynamic components that you don't think the 3.25 changes will cover, we'll see what we can come up with 🙂

**Comment by boris-petrov** [comment]
Thanks for taking the time to look into this.

> if you've got other places you're direct-invoking dynamic components that you don't think the 3.25 changes will cover

Well, my use-case is very specific and *very* dynamic (I compile templates on the backend and send them to the frontend where I register them as template-only components and then invoke them dynamically - and I don't know in advance how many templates there are, how they look like or anything at all) so the changes in 3.25 don't help me in any way, but this is a problem of mine and I've discussed this with pzuraq and he's been telling me that things will change and I have to adapt. So if you say that's a `wontfix`, I agree with you (and with your reasoning). :) No need to keep the issue open I guess.

Thanks again for the time! :)
